File: alarm_node/src/utils/encryption.ts
import * as crypto from 'crypto';
import logger from './logger';

/**
 * Encryption utility for sensitive configuration data
 * Uses AES-256-GCM for encryption
 */

const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 16;
const SALT_LENGTH = 64;

/**
 * Get encryption key from environment variable
 * If not set, generates a warning and uses a default (NOT SECURE FOR PRODUCTION)
 */
function getEncryptionKey(): Buffer {
  const key = process.env.CONFIG_ENCRYPTION_KEY;
  
  if (!key) {
    logger.warn('CONFIG_ENCRYPTION_KEY not set! Using insecure default key. SET THIS IN PRODUCTION!');
    // Default key for development only - DO NOT USE IN PRODUCTION
    return crypto.scryptSync('default-dev-key-change-in-production', 'salt', 32);
  }
  
  // Derive a proper 32-byte key from the provided key
  return crypto.scryptSync(key, 'encryption-salt', 32);
}

/**
 * Encrypt a plaintext string
 * Returns base64-encoded encrypted data in format: iv:authTag:encrypted
 */
export function encrypt(plaintext: string): string {
  try {
    if (!plaintext) {
      return '';
    }
    
    const key = getEncryptionKey();
    const iv = crypto.randomBytes(IV_LENGTH);
    const cipher = crypto.createCipheriv(ALGORITHM, key, iv);
    
    let encrypted = cipher.update(plaintext, 'utf8', 'base64');
    encrypted += cipher.final('base64');
    
    const authTag = cipher.getAuthTag();
    
    // Format: iv:authTag:encrypted (all base64)
    return `${iv.toString('base64')}:${authTag.toString('base64')}:${encrypted}`;
  } catch (error: any) {
    logger.error('Encryption error:', error);
    throw new Error(`Failed to encrypt data: ${error.message}`);
  }
}

/**
 * Decrypt an encrypted string
 * Expects format: iv:authTag:encrypted (all base64)
 */
export function decrypt(encryptedData: string): string {
  try {
    if (!encryptedData) {
      return '';
    }
    
    // Parse the encrypted data
    const parts = encryptedData.split(':');
    if (parts.length !== 3) {
      throw new Error('Invalid encrypted data format');
    }
    
    const [ivBase64, authTagBase64, encrypted] = parts;
    
    const key = getEncryptionKey();
    const iv = Buffer.from(ivBase64, 'base64');
    const authTag = Buffer.from(authTagBase64, 'base64');
    
    const decipher = crypto.createDecipheriv(ALGORITHM, key, iv);
    decipher.setAuthTag(authTag);
    
    let decrypted = decipher.update(encrypted, 'base64', 'utf8');
    decrypted += decipher.final('utf8');
    
    return decrypted;
  } catch (error: any) {
    logger.error('Decryption error:', error);
    throw new Error(`Failed to decrypt data: ${error.message}`);
  }
}

/**
 * Check if a string appears to be encrypted
 * (Simple check: has the format iv:authTag:encrypted)
 */
export function isEncrypted(data: string): boolean {
  if (!data) {
    return false;
  }
  
  const parts = data.split(':');
  if (parts.length !== 3) {
    return false;
  }
  
  // Check if all parts are valid base64
  try {
    parts.forEach(part => Buffer.from(part, 'base64'));
    return true;
  } catch {
    return false;
  }
}

/**
 * Encrypt an object's sensitive fields
 * Modifies the object in place and returns it
 */
export function encryptObject<T extends Record<string, any>>(
  obj: T,
  sensitiveFields: string[]
): T {
  const result: any = { ...obj };
  
  for (const field of sensitiveFields) {
    if (result[field] && typeof result[field] === 'string' && !isEncrypted(result[field])) {
      result[field] = encrypt(result[field]);
    }
  }
  
  return result as T;
}

/**
 * Decrypt an object's sensitive fields
 * Modifies the object in place and returns it
 */
export function decryptObject<T extends Record<string, any>>(
  obj: T,
  sensitiveFields: string[]
): T {
  const result: any = { ...obj };
  
  for (const field of sensitiveFields) {
    if (result[field] && typeof result[field] === 'string' && isEncrypted(result[field])) {
      try {
        result[field] = decrypt(result[field]);
      } catch (error: any) {
        logger.error(`Failed to decrypt field '${field}':`, error);
        // Leave as is if decryption fails
      }
    }
  }
  
  return result as T;
}

/**
 * Generate a secure random encryption key (for initial setup)
 * Returns a base64-encoded string suitable for CONFIG_ENCRYPTION_KEY
 */
export function generateEncryptionKey(): string {
  return crypto.randomBytes(32).toString('base64');
}

/**
 * Hash a password (for storing in database)
 * Uses scrypt which is resistant to GPU attacks
 */
export async function hashPassword(password: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const salt = crypto.randomBytes(SALT_LENGTH).toString('hex');
    
    crypto.scrypt(password, salt, 64, (err, derivedKey) => {
      if (err) reject(err);
      resolve(`${salt}:${derivedKey.toString('hex')}`);
    });
  });
}

/**
 * Verify a password against a hash
 */
export async function verifyPassword(password: string, hash: string): Promise<boolean> {
  return new Promise((resolve, reject) => {
    const [salt, storedHash] = hash.split(':');
    
    crypto.scrypt(password, salt, 64, (err, derivedKey) => {
      if (err) reject(err);
      resolve(storedHash === derivedKey.toString('hex'));
    });
  });
}

export default {
  encrypt,
  decrypt,
  isEncrypted,
  encryptObject,
  decryptObject,
  generateEncryptionKey,
  hashPassword,
  verifyPassword,
};

File: sms_gateway_node/utils/encryption.py
"""
Encryption utility for sensitive configuration data
Compatible with Alarm Service encryption (alarm_node folder). AES-256-GCM.
"""
import os
import base64
import hashlib
import logging
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend

logger = logging.getLogger(__name__)


def _get_encryption_key() -> bytes:
    """
    Get encryption key from environment variable.
    Uses scrypt to derive a 32-byte key (compatible with Node.js crypto.scryptSync)
    
    IMPORTANT: Node.js uses different salts:
    - Default key (no CONFIG_ENCRYPTION_KEY): salt = 'salt'
    - With CONFIG_ENCRYPTION_KEY: salt = 'encryption-salt'
    """
    key = os.environ.get('CONFIG_ENCRYPTION_KEY', '')
    
    if not key:
        logger.warning('CONFIG_ENCRYPTION_KEY not set! Using insecure default key.')
        key = 'default-dev-key-change-in-production'
        salt = b'salt'  # Node.js uses 'salt' for default key
    else:
        salt = b'encryption-salt'  # Node.js uses 'encryption-salt' when key is set
    
    # Derive key using scrypt (compatible with Node.js crypto.scryptSync)
    # Parameters match Node.js defaults: N=16384, r=8, p=1
    derived = hashlib.scrypt(
        key.encode('utf-8'),
        salt=salt,
        n=16384,
        r=8,
        p=1,
        dklen=32
    )
    return derived


def decrypt(encrypted_data: str) -> str:
    """
    Decrypt an encrypted string.
    Expects format: iv:authTag:encrypted (all base64)
    Compatible with Alarm Service encryption (alarm_node folder).
    """
    if not encrypted_data:
        return ''
    
    try:
        # Parse the encrypted data
        parts = encrypted_data.split(':')
        if len(parts) != 3:
            logger.warning('Invalid encrypted data format, returning as-is')
            return encrypted_data  # Return as-is if not encrypted
        
        iv_b64, auth_tag_b64, encrypted_b64 = parts
        
        # Decode from base64
        iv = base64.b64decode(iv_b64)
        auth_tag = base64.b64decode(auth_tag_b64)
        encrypted = base64.b64decode(encrypted_b64)
        
        # Get the key
        key = _get_encryption_key()
        
        # Decrypt using AES-256-GCM
        cipher = Cipher(
            algorithms.AES(key),
            modes.GCM(iv, auth_tag),
            backend=default_backend()
        )
        decryptor = cipher.decryptor()
        decrypted = decryptor.update(encrypted) + decryptor.finalize()
        
        return decrypted.decode('utf-8')
    
    except Exception as e:
        logger.error(f'Decryption error: {type(e).__name__}: {e}')
        import traceback
        logger.debug(f'Decryption traceback: {traceback.format_exc()}')
        # Return as-is if decryption fails (might be plaintext)
        return encrypted_data


def is_encrypted(data: str) -> bool:
    """
    Check if a string appears to be encrypted.
    (Simple check: has the format iv:authTag:encrypted)
    """
    if not data:
        return False
    
    parts = data.split(':')
    if len(parts) != 3:
        return False
    
    # Check if all parts are valid base64
    try:
        for part in parts:
            base64.b64decode(part)
        return True
    except Exception:
        return False
